src/embed/connection.rs

```diff
@@ -505,48 +513,41 @@ async fn do_scan(
     read: db::Read<'_>,
     req: ScanRequest,
     req_raw: JsValue,
+    lc: LogContext,
 ) -> Result<ScanResponse, ScanError> {
     let receiver: Function = Reflect::get(&req_raw, &JsValue::from_str("receiver"))
         .map_err(|_| ScanError::MissingReceiver)?
         .dyn_into()
         .map_err(|_| ScanError::InvalidReceiver)?;
 
-    use crate::prolly;
-
-    let using_index = req.opts.index_name.is_some();
-    read.scan(req.opts, |pe: prolly::Entry<'_>| {
-        // TODO This is the quick and dirty way to return primary and secondary
-        // index keys to the caller. The right way is probably to introduce
-        // a db::read::ScanResult that has a key, secondary_key, and val, and
-        // return those instead of prolly entries at the db::Read::scan level.
-        // I started to do this but hit lifetime issues and backed out.
-        //
-        // An alternative is to do it one level lower, at the scan::scan() level.
-        // Since scan::scan() doesn't know what kind of map it is scanning we
-        // would need to have a way for it to decode prolly map keys into
-        // indexed keys or regular ones, probably by adding a type byte to regular
-        // map keys (index map keys already have a type byte). This doesn't feel
-        // right tho -- scan::scan() should probably not know about indexes at all.
-        let primary_key: JsValue;
-        let secondary_key: JsValue;
-        if using_index {
-            // TODO the unwrap()s below are really unfortunate
-            let (secondary, primary) = db::index::decode_index_key(pe.key).unwrap();
-            primary_key = JsValue::from_str(std::str::from_utf8(&primary).unwrap());
-            secondary_key = JsValue::from_str(&secondary);
-        } else {
-            primary_key = JsValue::from_str(std::str::from_utf8(pe.key).unwrap());
-            secondary_key = JsValue::null();
+    read.scan(req.opts, |sr: db::ScanResult<'_>| {
+        match sr {
+            db::ScanResult::Error(e) => error!(lc, "Error returning scan result: {:?}", e),
```

Comment:
Is this throwing in JS? I think it must. Just logging the error is going to result in a JS array of empty entries/items.

Reply:
Strawman proposal. Pass in a thrower function that gets called with a string.

Reply:
I'm not sure why it would throw in js? error!() should just log an error. the scan will continue.

Reply:
to be clear what happens is that when we get a result error instead of an item we log an error, it does not invoke the callback, and it continues on to the next item.

Reply:
My point is that we should throw, or tell the bindings about the error. Just logging the error leads to invalid state.